### ISSUE TITLE
Quest: First-Class Component Templates

### ISSUE THREAD

**Issue by chriskrycho** [emberjs/ember.js#20062]
The path to making [RFC #0779: First-Class Component Templates][fccts] *stable* and *recommended* for end users.

[fccts]: https://emberjs.github.io/rfcs/0779-first-class-component-templates.html

<details><summary>What even is this?</summary>

## Background

[RFC #0779: First-Class Component Templates][fccts] picked `<template>` tags as the the ability for Ember (and Glimmer) developers to write components with `template` tags embedded in their JS, like so:

```js
import Component from '@glimmer/component';

const Button = <template>
  <button
    type='button'
    class="my-button"
    {{on "click" @onClick}}
  >
    {{@label}}
  </button>
</template>

export default class Player extends Component {
  @tracked isPlaying = false;

  play = () => (this.isPlaying = true);
  pause = () => (this.isPlaying = false);

  <template>
    <audio src={{@srcUrl}} {{playWhen this.isPlaying}} />

    <Button @label="Play" @onClick={{this.play}} />
    <Button @label="Pause" @onClick={{this.pause}} />
  </template>
}
```

This is not only a general win for ease of working with Ember: it is also a key part of our TypeScript story, because it makes things like go-to-definition, documentation hovers, autocomplete, auto-import, etc. all *much* easier to manage. So: what do we need to do to finish unblocking its adoption?

</details>

## The Road to Stable

See ember-template-imports/ember-template-imports#35 for details on each part of this.

- [ ] Prettier support
- [ ] ESLint support
- [ ] ember-template-lint integration

## The Road to Recommended

- [ ] Finish and publish spec
- [ ] Make any spec-driven updates to `ember-template-imports`
- [ ] Integrate `ember-template-imports` into Glimmer, with Ember being a customer

**Comment by MelSumner** [comment]
Needs prioritizing: - Needs prioritizing: https://github.com/emberjs/rfcs/issues/816

**Comment by chriskrycho** [comment]
- Thanks to @ventuno, we have [ember-template-lint](https://github.com/ember-template-lint/ember-template-lint/pull/2483) support!
- Thanks to @gitKrystan, we have [Prettier support](https://github.com/gitKrystan/prettier-plugin-ember-template-tag)!
- Thanks to @NullVoxPopuli, we have [ESLint support](https://github.com/ember-cli/eslint-plugin-ember/pull/1395)!

I’ll make sure we follow up on this in one of the next few Framework Core team meetings—this is amazing progress, and represents a huge step toward stabilizing this and eventually making it *the* recommended way of writing Ember templates.

(Keep your eyes open for code-mods coming soon™!)

**Comment by ef4** [comment]
A status update here is that 
 - we merged https://github.com/emberjs/rfcs/blob/master/text/0931-template-compiler-api.md which stabilized the plain-javascript representation of template tag
 - the build-time implementation of that RFC shipped in https://github.com/embroider-build/content-tag/ for going from template tag to the low-level format. This implementation fixes many outstanding bugs in the original ember-template-imports implementation, which didn't have the benefit of a full javascript parser.
 - and babel-plugin-ember-template-compilation@2.2.0 shipped with the ability to take the low-level format to runnable code.

**Comment by wagenet** [comment]
@ef4 Do we have an updated checklist on this?

**Comment by ef4** [comment]
- [ ] publish a version of ember-cli-htmlbars that depends on a minimum babel-plugin-ember-template-compilation version of 2.2.0 and that adds the `@babel/plugin-transform-class-static-block` plugin whenever adding the other class-feature plugins that are implied by legacy decorators.
 - [ ]  publish a version of embroider that depends on a minimum babel-plugin-ember-template-compilation version of 2.2.0 and that adds the `@babel/plugin-transform-class-static-block` plugin whenever adding the other class-feature plugins that are implied by legacy decorators.
 - [ ] publish a new ember-template-imports that uses content-tag to do preprocessing and does nothing on the babel side, documenting that it requires the above mentioned ember-cli-htmlbars or embroider version.
 - [ ] document clearly how to get full editor support (syntax highlighting, linters, typescript) when using template tag in (at least) VSCode.
 - [ ] ensure that the blueprints support prettier / eslint / etc of GJS/GTS out of the box.
 - [ ] update guides to use `<template>`

**Comment by NullVoxPopuli** [comment]
couple clarifying questions:

> document 

where to do this?
linting should be automatic provided new enough eslint-plugin-ember and ember-template-lint.

> update guides to use

pre-polaris? if so, this would be a good thing to start soon, given all the implementation stuff is done? (this is a really big _given_.)
there is a bunch of infra-work for the ember-family-of-sites blocking this, afaik (empress, guide-maker, highlighting, (my prism PRs, etc))

**Comment by NullVoxPopuli** [comment]
## Remaining todos

Required

- [ ] https://github.com/emberjs/rfcs/pull/933
- [ ] https://github.com/emberjs/rfcs/pull/991
- [ ] https://github.com/embroider-build/content-tag/issues/39
- [ ] https://github.com/embroider-build/content-tag/issues/61
  - [ ] https://github.com/embroider-build/content-tag/pull/29
  - [ ] https://github.com/embroider-build/content-tag/pull/56
  - [ ] https://github.com/embroider-build/content-tag/pull/53
- [ ] https://github.com/emberjs/rfcs/pull/976 (since tsserver doesn't work on gjs/gts)
  - Needs Fixed/Resolved
  - [ ] https://github.com/typed-ember/glint/issues/694
  - [ ] https://github.com/typed-ember/glint/issues/692
  - [ ] https://github.com/typed-ember/glint/issues/676
  - [ ] https://github.com/typed-ember/glint/issues/672
  - [ ] https://github.com/typed-ember/glint/issues/667
  - [ ] https://github.com/typed-ember/glint/issues/661
  - [ ] https://github.com/typed-ember/glint/issues/649
  - [ ] https://github.com/typed-ember/glint/issues/645
  - [ ] https://github.com/typed-ember/glint/issues/633
  - [ ] https://github.com/typed-ember/glint/issues/628
  - [ ] https://github.com/typed-ember/glint/issues/626
  - [ ] https://github.com/typed-ember/glint/issues/617
  - [ ] https://github.com/typed-ember/glint/issues/614
  - [ ] https://github.com/typed-ember/glint/issues/609
  - [ ] https://github.com/typed-ember/glint/issues/604
  - [ ] https://github.com/typed-ember/glint/issues/599
  - [ ] https://github.com/typed-ember/glint/issues/589
  - [ ] https://github.com/typed-ember/glint/issues/575
  - [ ] https://github.com/typed-ember/glint/issues/506
  - [ ] https://github.com/typed-ember/glint/issues/271
  - [ ] https://github.com/typed-ember/glint/issues/121


Nice to have
- [ ] https://github.com/emberjs/rfcs/pull/998
- [ ] https://github.com/typed-ember/glint/issues/597
- [ ] https://github.com/typed-ember/glint/issues/387
